internal/provider/resource_access_offer.go

```diff
@@ -0,0 +1,436 @@
+	state.Access = types.StringValue(access)
+	state.OfferURL = types.StringValue(offerURL)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (a *accessOfferResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "read")
+		return
+	}
+	var plan accessOfferResourceOffer
+
+	// Get the plan
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Retrieve information from the plan
+	offerURL := plan.OfferURL.ValueString()
+	access := plan.Access.ValueString()
+	var planUsers []string
+	if err := plan.Users.ElementsAs(ctx, &planUsers, false); err != nil {
+		resp.Diagnostics.AddError("Plan error", fmt.Sprintf("Unable to read users from the plan, got error: %s", err))
+		return
+	}
+
+	// Get user/access info from Offer
+	response, err := a.client.Offers.ReadOffer(&juju.ReadOfferInput{
+		OfferURL: offerURL,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read offer, got error: %s", err))
+		return
+	}
+	a.trace(fmt.Sprintf("read juju offer %q", offerURL))
+	offerUsers := response.Users
+
+	// Check if a user was removed and delete access (revoke read)
+	// Create map to be used in the next step for adding new users
+	offerUserNames := make(map[string]struct{})
+	for _, offerUser := range offerUsers {
+		if offerUser.UserName == "everyone@external" || offerUser.UserName == "admin" {
+			continue
+		}
+		offerUserNames[offerUser.UserName] = struct{}{}
+		if !slices.Contains(planUsers, offerUser.UserName) {
+			a.trace(fmt.Sprintf("revoke user %q for offer %q", offerUser.UserName, offerURL))
+			err := a.client.Offers.RevokeOffer(&juju.GrantRevokeOfferInput{
+				User:     offerUser.UserName,
+				Access:   "read",
+				OfferURL: offerURL,
+			})
+			if err != nil {
+				resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update access offer resource, got error: %s", err))
+				return
+			}
+		}
+	}
+	// Check if is a new user and grant access
+	// If not, revoke all access and grant the right one
+	for _, planUser := range planUsers {
+		if _, ok := offerUserNames[planUser]; !ok {
+			// user has no access yet so we need to grant
+			a.trace(fmt.Sprintf("grant permission %q to user %q for offer %q", access, planUser, offerURL))
+			err := a.client.Offers.GrantOffer(&juju.GrantRevokeOfferInput{
+				User:     planUser,
+				Access:   access,
+				OfferURL: offerURL,
+			})
+			if err != nil {
+				resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access offer resource, got error: %s", err))
+				return
+			}
+			continue
+		}
+		// revoke read (remove all)
+		a.trace(fmt.Sprintf("revoke user %q for offer %q", planUser, offerURL))
+		err := a.client.Offers.RevokeOffer(&juju.GrantRevokeOfferInput{
```

Comment:
this might cause some interesting/unwanted side-effects. imagine a plan with multiple juju_access_offer resources

resource "juju_access_offer" "this_consume" {
  offer_url = "offer1"
  access = "consume"
  users  = ["user1", "user2", "user3"]
}

resource "juju_access_offer" "this_read" {
  offer_url = "offer1"
  access = "read"
  users  = ["user4", "user5"]
}
which grants user3 consume permission on the offer, i apply this terraform plan, then decide to move user3 to the read resource so that i get

resource "juju_access_offer" "this_consume" {
  offer_url = "offer1"
  access = "consume"
  users  = ["user1", "user2"]
}

resource "juju_access_offer" "this_read" {
  offer_url = "offer1"
  access = "read"
  users  = ["user3", "user4", "user5"]
}
now, depending on the order of execution, user3 might end up with no permissions, because updating the consume resource revokes user3's read access..

Reply:
For the new schema this no longer apply.